### PR TITLE
Sensors errors and integration with control

### PR DESCRIPTION
**Pull request by cua-cua**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,7 @@ __pycache__/
 *.out
 *.stylecheck
 *.swp
+config.h
```
Comment (Peque) [inline .gitignore]: I would not add the file to the `.gitignore` file.

File: src/control.c [hunk 2]
```diff
@@ -19,11 +19,19 @@ static volatile float kp_linear = 1600.;
...
+static volatile float ki_angular_side = 40.;
+static volatile float side_sensors_error_factor = 50.;
+static volatile float front_sensors_error_factor = 20.;
 
 static volatile int32_t pwm_left;
 static volatile int32_t pwm_right;
 
 static volatile bool collision_detected_signal;
+static volatile bool side_sensors_control_enable;
+static volatile bool front_sensors_control_enable;
```
Comment (Peque) [inline src/control.c]: I would say "enabled" rather than "enable".

File: src/detection.c [hunk 3]
```diff
@@ -226,3 +218,39 @@ float get_side_right_distance(void)
 {
 	return distance[SENSOR_SIDE_RIGHT_ID];
 }
+
+/**
+ * @brief Calculate and return the side sensors error.
+ *
+ * Taking into account that the walls are parallel to the robot, this function
+ * returns the distance that the robot is moved from the center parallel line
+ * of the maze.
```
Comment (Peque) [inline src/detection.c]: "center of the corridor"? instead of "center parallel line of the maze"?

File: src/detection.c [hunk 4]
```diff
@@ -226,3 +218,39 @@ float get_side_right_distance(void)
...
+		    distance[SENSOR_SIDE_RIGHT_ID] - MIDDLE_MAZE_DISTANCE;
+	} else if ((distance[SENSOR_SIDE_RIGHT_ID] > MIDDLE_MAZE_DISTANCE) &&
+		   (distance[SENSOR_SIDE_LEFT_ID] < MIDDLE_MAZE_DISTANCE)) {
+		sensors_error =
+		    MIDDLE_MAZE_DISTANCE - distance[SENSOR_SIDE_LEFT_ID];
+	} else {
+		sensors_error = 0;
+	}
+	return sensors_error;
+}
```
Comment (Peque) [inline src/detection.c]: I'd use the `if (...) return;` syntax here, but I guess that's a matter of preference... :stuck_out_tongue_winking_eye: 

See for example:

```c
float get_side_sensors_error(void)
{
	if ((distance[SENSOR_SIDE_LEFT_ID] > MIDDLE_MAZE_DISTANCE) &&
	    (distance[SENSOR_SIDE_RIGHT_ID] < MIDDLE_MAZE_DISTANCE))
		return distance[SENSOR_SIDE_RIGHT_ID] - MIDDLE_MAZE_DISTANCE;
	if ((distance[SENSOR_SIDE_RIGHT_ID] > MIDDLE_MAZE_DISTANCE) &&
		   (distance[SENSOR_SIDE_LEFT_ID] < MIDDLE_MAZE_DISTANCE))
		return MIDDLE_MAZE_DISTANCE - distance[SENSOR_SIDE_LEFT_ID];
	return 0;
}
```

File: src/logging.c [hunk 5]
```diff
@@ -87,3 +97,46 @@ void log_sensors_distance(void)
...
+	uint16_t off[NUM_SENSOR];
+	uint16_t on[NUM_SENSOR];
+
+	get_sensors_raw(off, on);
+
+	LOG_INFO("OFF-ON,%d,%d,%d,%d,%d,%d,%d,%d", off[SENSOR_SIDE_LEFT_ID],
+		 off[SENSOR_SIDE_RIGHT_ID], off[SENSOR_FRONT_LEFT_ID],
+		 off[SENSOR_FRONT_RIGHT_ID], on[SENSOR_SIDE_LEFT_ID],
+		 on[SENSOR_SIDE_RIGHT_ID], on[SENSOR_FRONT_LEFT_ID],
+		 on[SENSOR_FRONT_RIGHT_ID]);
```
Comment (Peque) [inline src/logging.c]: I would omit the "OFF-ON" column here.

File: src/logging.c [hunk 6]
```diff
@@ -87,3 +97,46 @@ void log_sensors_distance(void)
...
+ * @brief Log front sensors distances and error.
+ */
+void log_front_sensors_error(void)
+{
+	float sensors_error = get_front_sensors_error();
+	float fl_dist = get_front_left_distance();
+	float fr_dist = get_front_right_distance();
+
+	LOG_INFO("front_sensors_error:,%f,distances:%f,%f", sensors_error,
+		 fl_dist, fr_dist);
```
Comment (Peque) [inline src/logging.c]: This is not easily parseable. Either use "%f,%f,%f" or, if you want to include the variable names, then use something like "{"name":%f,...}" (see the `control_variables` above).

File: src/logging.c [hunk 7]
```diff
@@ -87,3 +97,46 @@ void log_sensors_distance(void)
...
+ * @brief Log side sensors distances and error.
+ */
+void log_side_sensors_error(void)
+{
+	float sensors_error = get_side_sensors_error();
+	float sl_dist = get_side_left_distance();
+	float sr_dist = get_side_right_distance();
+
+	LOG_INFO("side_sensors_error:,%f,distances:%f,%f", sensors_error,
+		 sl_dist, sr_dist);
```
Comment (Peque) [inline src/logging.c]: Same comment as above (choose between "%f,%f,%f" or the JSON parseable dictionary).

File: src/control.c (hunk 2 above)
Comment (cua-cua) [reply src/control.c]: Fixed, they are not volatile either.

File: src/detection.c (hunk 3 above)
Comment (cua-cua) [reply src/detection.c]: good

File: src/logging.c (hunk 5 above)
Comment (cua-cua) [reply src/logging.c]: I wrote it just as a comment to read the values better on terminal.

File: src/detection.c (hunk 4 above)
Comment (cua-cua) [reply src/detection.c]: I prefer the else if, else. 😇

File: src/control.c (hunk 2 above)
Comment (Peque) [reply src/control.c]: I think they are. :blush:

**Pull request merged**